### PR TITLE
fix: make transform.sh patches fail loudly

### PR DESCRIPTION
**Pull request by NikolayS**
# Build-pipeline integrity fixes (transform.sh)

Addresses findings C1, C2, C3, C5, C8 of #283.

### C1 — silent no-op sed patches (medium)
The critical correctness patches (batch_event_sql xid8 quoting, `batch_retry` `NULL::int8 → NULL::xid8`, `upgrade_schema` `return 0 → return cnt`, sqltriga header fix, annotation seds) were plain `sedi` calls: sed exits 0 whether or not it matched, and the script printed `PASS: ...` unconditionally. If upstream PgQ reworded an anchored line, the build would report success while the generated `batch_event_sql` emits unquoted txids → runtime `operator does not exist: xid8 = integer` on every batch fetch.

**Fix:** new `sedi_must` helper (cmp before/after; exits 1 with the failing expression and file when nothing changed). Every must-match `sedi` now goes through it. The `upgrade_schema` role-grant and `insert_event_raw` guard awk patches now count their substitutions/injections and fail on a count mismatch, matching the standard already set by the get_batch_cursor patch.

### C2 — missing source files only warned (medium)
- Transform loop: `WARNING ... continue` → hard error (`exit 1`).
- `FUNCTION_FILES` assembly loop: `WARNING` → hard error.
- `pgque-api` files: previously **silently skipped with no message** → hard error; a missing `sql/pgque-api/` directory is also a hard error.
- `pgque-additions` files: explicit hard error with a clear message.

A submodule bump that renames/drops a file can no longer ship a truncated install file that passes the CI drift check after regenerate-and-commit.

### C3 — vacuous pg_notify check + unverified ticker injection (medium)
- (a) The ticker pg_notify awk injection now counts both anchors (`return i_tick_id;`, `return currval(q.queue_tick_seq);`) and fails the build unless each fired exactly once. Same for the create_queue length-check injection.
- (b) The post-assembly check `grep -q 'pg_notify' "$INSTALL_FILE"` was vacuous — Section 6 (hardening.sql) contains its own pg_notify, so it passed even if BOTH ticker injections missed. Replaced with an awk that extracts each tick-emitting `pgque.ticker` definition from the generated file and requires `perform pg_notify(` inside each body.

### C5 — bogus annotation on hardening.sql's pg_notify line (low, changes generated files)
The `-- PgQue transformation: LISTEN/NOTIFY wakeup (not in original PgQ)` annotation sed ran against the whole assembled file and also tagged the pg_notify line inside hardening.sql's external-ticker override — a line that was never in PgQ — making the embedded chunk differ from `sql/pgque-additions/hardening.sql`.

**Fix:** the annotation block now runs before Sections 6/7 are appended, so only the transformed-PgQ portion is annotated. Regenerated files: the **only** change to `sql/pgque.sql` and `sql/pgque-tle.sql` is that line losing the bogus trailing comment, making the embedded hardening chunk byte-identical to its source (verified with a chunk diff, see below).

### C8 — hardcoded header version (low)
The generated header hardcoded `-- Version: 0.2.0` while the pg_tle registration version was extracted dynamically from `pgque.version()` in `lifecycle.sql`. A version bump in lifecycle.sql would leave the header stale with no check.

**Fix:** the version is extracted once, early (same anchored awk + semver validation), into `PGQUE_VERSION` and reused for both the header and the pg_tle registration. Current version is 0.2.0, so regeneration does not change the version text.

### 1. Build on this branch: exit 0, expected diff only

```
$ bash build/transform.sh; echo exit=$?
...
=== ASSEMBLY COMPLETE — ALL CHECKS PASSED ===
=== pg_tle PACKAGING COMPLETE ===
exit=0

$ git diff --stat   # (before commit)
 build/transform.sh | 361 ++++++++++++++++++++++++++++++++++++++---------------
 sql/pgque-tle.sql  |   2 +-
 sql/pgque.sql      |   2 +-
```

The full generated-file diff (the single C5 annotation line, once per file):

```
--- a/sql/pgque.sql
+++ b/sql/pgque.sql
@@ -5187,7 +5187,7 @@ begin
-    perform pg_notify('pgque_' || i_queue_name, i_tick_id::text); -- PgQue transformation: LISTEN/NOTIFY wakeup (not in original PgQ)
+    perform pg_notify('pgque_' || i_queue_name, i_tick_id::text);
--- a/sql/pgque-tle.sql
+++ b/sql/pgque-tle.sql
@@ -5275,7 +5275,7 @@ begin
-    perform pg_notify('pgque_' || i_queue_name, i_tick_id::text); -- PgQue transformation: LISTEN/NOTIFY wakeup (not in original PgQ)
+    perform pg_notify('pgque_' || i_queue_name, i_tick_id::text);
```

### 2. Negative tests (submodule restored with `git -C pgq checkout .` after each; nothing committed)

**C1 — reword the batch_event_sql anchor line:**

```
$ sed -i "s/|| ' and ev.ev_txid >= ' || batch.tx_start::text/|| ' and ev.ev_txid>=' || batch.tx_start::text/" \
    pgq/functions/pgq.batch_event_sql.sql
$ bash build/transform.sh; echo exit=$?
...
FAIL: batch_event_sql xid8 quoting (tx_start lower bound) did not apply
      sed expression matched nothing in .../build/output/functions/pgque.batch_event_sql.sql:
      s/|| ' and ev.ev_txid >= ' || batch.tx_start::text/|| ' and ev.ev_txid >= ''' || batch.tx_start::text || '''::xid8'/
      Upstream PgQ likely reworded the anchor line; update
      build/transform.sh to match the new source.
exit=1
```

**C2 — delete an expected source file:**

```
$ rm pgq/functions/pgq.batch_retry.sql
$ bash build/transform.sh; echo exit=$?
ERROR: source file not found: functions/pgq.batch_retry.sql
       A submodule bump may have renamed or dropped it; update
       SOURCE_FILES in build/transform.sh before rebuilding.
exit=1
```

**C2 — remove a pgque-api file (previously silently skipped):**

```
$ mv sql/pgque-api/send.sql /tmp/ && bash build/transform.sh; echo exit=$?
ERROR: pgque-api file not found: pgque-api/send.sql
       Shipping without it would truncate the install file.
exit=1
```

**C3 — break the ticker-injection anchor indentation:**

```
$ sed -i 's/^    return i_tick_id;$/        return i_tick_id;/' pgq/functions/pgq.ticker.sql
$ bash build/transform.sh; echo exit=$?
ERROR: ticker pg_notify injection anchors matched 0/1 times (expected 1/1)
FAIL: pg_notify injection into ticker did not happen; anchor lines
      in pgq/functions/pgq.ticker.sql may have changed upstream.
exit=1
```

**C3b — the new post-assembly check is non-vacuous.** Against a doctored copy of `sql/pgque.sql` with the two injected ticker pg_notify lines stripped (hardening.sql's own pg_notify still present — this fooled the old whole-file grep, `grep -c 'perform pg_notify'` = 1):

```
ERROR: pg_notify missing from ticker bodies (4-arg: 0/1, 1-arg: 0/1)
check exit=1
```

**C8 — version propagation (temporary bump of lifecycle.sql to 9.9.9, then restored):**

```
$ grep -n -- '-- Version:' sql/pgque.sql | head -1
2:-- Version: 9.9.9
$ grep -n '9.9.9' sql/pgque-tle.sql | head -1
71:    if existing_version = '9.9.9' then
```

### 3. Fresh install + full test suite

```
$ createdb pgque_build_oqxpbr
$ psql -X -v ON_ERROR_STOP=1 -d pgque_build_oqxpbr -f sql/pgque.sql      # exit 0
$ psql -X -v ON_ERROR_STOP=1 -d pgque_build_oqxpbr -f tests/run_all.sql  # exit 0
...
=== ALL TESTS PASSED ===
```

### 4. Hardening chunk byte-identity

```
$ diff <(extracted pgque-additions/hardening.sql chunk from sql/pgque.sql) sql/pgque-additions/hardening.sql
CHUNK BYTE-IDENTICAL
```

## Manual verification command

```
bash build/transform.sh && git diff --exit-code sql/ \
  && createdb pgque_scratch && psql -X -v ON_ERROR_STOP=1 -d pgque_scratch -f sql/pgque.sql \
  && psql -X -v ON_ERROR_STOP=1 -d pgque_scratch -f tests/run_all.sql
```

Note: other branches from #283 also regenerate `sql/pgque.sql` / `sql/pgque-tle.sql`; merge conflicts there are expected and should be resolved by re-running `build/transform.sh`. The TLE header wording about client compatibility (finding C7) is intentionally untouched here.

https://claude.ai/code/session_01KAaEGkQZmey1D1xCsVGmqv

---
_Generated by [Claude Code](https://claude.ai/code/session_01KAaEGkQZmey1D1xCsVGmqv)_

### CODE REVIEW

**Comment by NikolayS** [comment]
## samorev Code Review Report

- **PR:** NikolayS/PgQue#292 - fix: make transform.sh patches fail loudly
- **Author:** @NikolayS
- **AI-Assisted:** Unknown

| Pipeline | Coverage |
|----------|----------|
| PASS | Not reported |

---

### BLOCKING ISSUES (1)

**HIGH** `MR/PR state` - Review target is draft
> The review target is still marked as draft.
> **Fix:** Mark it ready for review before merge.

---

### Summary

| Area | Findings | Potential | Filtered |
|------|----------|-----------|----------|
| CI/Pipeline | 0 | 0 | 0 |
| Security | 0 | 0 | 0 |
| Bugs | 0 | 0 | 0 |
| Tests | 0 | 0 | 0 |
| Guidelines | 0 | 0 | 0 |
| Docs | 0 | 0 | 0 |
| Metadata | 1 | 0 | 0 |

Note:
- **Findings**: High-confidence issues (8-10/10) - blocking or non-blocking per severity
- **Potential**: Medium-confidence issues (4-7/10) - review manually
- **Filtered**: Low-confidence issues (0-3/10) - excluded as likely false positives

<details>
<summary>Review metadata</summary>

```text
provider=github
kind=pr
project=NikolayS/PgQue
number=292
target=github:NikolayS/PgQue#292
state=OPEN
draft=true
diff_lines=576
diff_added=264
diff_removed=100
diff_bytes=24759
comments_count=0
commits_count=1
ci_status=success
ci_summary=total=14 success=14 failure=0 pending=0 other=0
prompt=.claude/commands/review-mr.md
blocking=false
posted_by=gh
no_comment=false
live_posting=posted
```

</details>

---
*samorev-assisted review (AI analysis by [Tanya301/samorev](https://github.com/Tanya301/samorev))*

**Comment by NikolayS** [comment]
## Post-merge deep review — PR #292 (`fix: make transform.sh patches fail loudly`)

Reviewed the full diff and verified empirically against a scratch clone at the merge commit
`cd3b71d` (submodule pin `d23425f`), Postgres 18. **No blocking findings.** The fail-loudly
mechanism is sound and CI-enforced. Details and a few minor notes below.

### Empirical verification (all green)

| Check | Result |
|---|---|
| Unmodified `bash build/transform.sh` | exit 0, `git diff sql/` empty → **byte-identical** regeneration |
| Fresh install `sql/pgque.sql` on PG18 | exit 0 |
| `tests/run_all.sql` | `=== ALL TESTS PASSED ===` |
| Reword `batch_event_sql` anchor | exit 1, `FAIL: batch_event_sql xid8 quoting (tx_start lower bound) did not apply` |
| Delete source `pgq.batch_retry.sql` | exit 1, `ERROR: source file not found` |
| Break ticker `return i_tick_id;` indentation | exit 1, `ERROR: ticker pg_notify injection anchors matched 0/1` |
| Bump `lifecycle.sql` to `9.9.9` | header + pg_tle both propagate `9.9.9` (no drift) |

### (a) Correctness of fail-loudly — coverage is complete

Every correctness-critical patch site now fails on a no-op. I grepped for any plain `sedi`/`awk`
left unguarded:
- The only bare `sedi` outside `sedi_must` is the `get_queue_info` promotion (line 716) — guarded by
  the `defcount -ne 2` check immediately after (720–725).
- The `get_batch_cursor` awk has its own `header_done → exit 1` guard (697).
- All bulk mechanical transforms (schema rename, txid, search_path, `queue_per_tx_limit`, oids,
  hooks, magic columns) are covered by the aggregate "nothing-remains" self-verification greps
  (343–428).

No missed site.

### (b) Shell robustness — sound

- `sedi_must` detects no-op via `cp` + `cmp -s` (not sed exit code, which is always 0). Correct
  approach for "sed doesn't report no-match".
- Cross-platform `sedi` (BSD `-i ''` vs GNU `-i`) is preserved; `cmp`/`awk`/`grep -c`/`wc -l` are
  portable. Verified the negative paths on macOS BSD sed; CI runs the same script on Linux.
- `.pre` temp files are always `rm -f`'d, and `build/output/` is gitignored, so no leak into commits.
- The injection awks fail the pipe (`if ! awk … ; then rm tmp; exit 1`) — a failed injection cannot
  leave a half-written `.tmp` in place.

### (c) Generated-file diff is genuinely cosmetic

Confirmed by re-running transform.sh at the merge commit: the committed `sql/pgque.sql` /
`sql/pgque-tle.sql` reproduce byte-identically, and the only diff vs the parent is the single C5
annotation-comment removal per file (the LISTEN/NOTIFY tag that was wrongly applied to
hardening.sql's own `pg_notify`). Matches the PR description.

### (d) CI coverage — adequate, one gap

`bash build/transform.sh` runs as a gating step in every `ci.yml` job, and a dedicated **drift check**
(`git diff --exit-code -- sql/pgque.sql sql/pgque-tle.sql`, lines 676–689) fails the job on
regeneration drift. Because the script now `exit 1`s on any silent no-op, a future submodule bump
that breaks an anchor turns CI red at the build step. That is the real protective outcome and it is
covered.

---

### Minor / nits (non-blocking, optional follow-ups)

1. **[minor] No regression test exercises the fail-loud paths.** Protection is real (CI gates on the
   script's exit code), but nothing in `tests/` deliberately breaks an anchor to assert `exit 1`, so a
   future refactor of `sedi_must` / the injection awks could regress silently. A tiny shell test
   (break one anchor in a temp copy of the submodule, assert non-zero) would lock in the behavior.

2. **[nit] `sedi_must` verifies "something changed", not "the intended line / exact count."** `cmp`
   passes if any byte differs, so an expression that matched an unintended line would still pass. Not
   a real risk here — the anchors are highly specific, and the functional injections (ticker,
   create_queue, insert_event_raw guard) additionally carry count-based awk checks and the
   post-assembly ticker-body check. Noting only so the limitation is understood if these expressions
   are ever loosened.

3. **[nit] Version regex rejects SemVer build metadata.**
   `^[0-9]+\.[0-9]+\.[0-9]+(-[A-Za-z0-9.]+)?$` accepts pre-release (`-rc.1`) but not `+build`. Only
   bites if the project ever adopts `+`-metadata versions; harmless today (0.2.0).

4. **[nit, pre-existing / out of scope] Rename-vs-remove blind spot in aggregate greps.** The
   "nothing-remains" checks (e.g. `pgq_node`/`londiste` removal) would silently pass if upstream
   *renamed* those hooks rather than kept the old names — the removal sed no-ops and the grep finds
   nothing. Not introduced by this PR; flagged for awareness.

Bottom line: PR #292 does what it claims, regenerates byte-identically, installs and tests clean on
PG18, and is genuinely CI-enforced. Safe as merged. The items above are optional hardening, not
fixes.

**Comment by NikolayS** [comment]
## samorev Code Review Report

- **PR:** NikolayS/pgque#292 - fix: make transform.sh patches fail loudly
- **Author:** @NikolayS
- **AI-Assisted:** Yes

| Pipeline | Coverage |
|----------|----------|
| PASS | Not reported |

---

> This report supersedes the earlier `samorev review --fetch` comment on this PR. That comment was a CI/draft **gate check only** and performed no code analysis (it hardcoded zeros in every area). The results below come from a genuine per-area review of the diff, verified empirically against a scratch clone at the merge commit on PostgreSQL 18.

**No blocking issues.** The fail-loudly mechanism is sound and CI-enforced. One non-blocking suggestion and three lower-confidence notes below.

### What was verified

- **Fail-loud coverage is complete.** Every correctness-critical patch site fails on a no-op: the only bare `sedi` outside `sedi_must` (the `get_queue_info` promotion) is guarded by the immediately-following `defcount -ne 2` check; the `get_batch_cursor` awk has its own `header_done → exit 1`; bulk transforms are covered by the aggregate "nothing-remains" greps.
- **Negative paths fire correctly:** rewording the `batch_event_sql` anchor → `exit 1`; deleting a source file → `ERROR: source file not found`; breaking the ticker injection indentation → `ticker pg_notify injection anchors matched 0/1`.
- **Shell robustness:** `sedi_must` detects a no-op via `cp` + `cmp -s` (not sed's always-zero exit); cross-platform BSD/GNU `sedi` preserved; `.pre` temps always `rm -f`'d; failed injection awks fail the pipe and leave no half-written `.tmp`.
- **Genuinely cosmetic generated diff:** re-running `transform.sh` at the merge commit reproduces `sql/pgque.sql`/`sql/pgque-tle.sql` byte-identically; the only diff vs parent is the single stray annotation-comment removal per file.
- **CI-enforced:** `bash build/transform.sh` gates every `ci.yml` job and a dedicated drift check (`git diff --exit-code` on the generated files) fails on regeneration drift — a future submodule bump that breaks an anchor turns CI red at the build step.

---

### NON-BLOCKING (1)

**LOW** `build/transform.sh` - no regression test exercises the fail-loud paths
> Protection is real (CI gates on the script's exit code), but nothing in `tests/` deliberately breaks an anchor to assert `exit 1`, so a future refactor of `sedi_must` / the injection awks could regress silently.
> **Suggestion:** Add a tiny shell test that breaks one anchor in a temp copy of the submodule and asserts non-zero exit, to lock in the behavior. Optional hardening.

---

### POTENTIAL ISSUES (2)

**LOW** `build/transform.sh` - `sedi_must` verifies "something changed", not the intended line/count *(confidence: 5/10)*
> `cmp` passes if any byte differs, so an expression matching an unintended line would still pass.
> **Suggestion:** Not a real risk today — anchors are highly specific and functional injections carry additional count-based awk checks. Noting the limitation in case these expressions are ever loosened.

**LOW** `build/transform.sh` - version regex rejects SemVer build metadata *(confidence: 4/10)*
> `^[0-9]+\.[0-9]+\.[0-9]+(-[A-Za-z0-9.]+)?$` accepts pre-release (`-rc.1`) but not `+build`.
> **Suggestion:** Only bites if the project ever adopts `+`-metadata versions; harmless at 0.2.0.

---

### Summary

| Area | Findings | Potential | Filtered |
|------|----------|-----------|----------|
| CI/Pipeline | 0 | 0 | 0 |
| Security | 0 | 0 | 0 |
| Bugs | 0 | 2 | 1 |
| Tests | 1 | 0 | 0 |
| Guidelines | 0 | 0 | 0 |
| Docs | 0 | 0 | 0 |
| Metadata | 0 | 0 | 0 |

Note:
- **Findings**: High-confidence issues (8-10/10) - blocking or non-blocking per severity
- **Potential**: Medium-confidence issues (4-7/10) - review manually
- **Filtered**: Low-confidence issues (0-3/10) - excluded as likely false positives

Filtered (1): rename-vs-remove blind spot in the aggregate "nothing-remains" greps (e.g. `pgq_node`/`londiste`) — a future upstream *rename* would silently pass. Pre-existing, not introduced by this PR; excluded as out of scope.

<details>
<summary>Review metadata</summary>

```text
provider=github
kind=pr
project=NikolayS/pgque
number=292
target=github:NikolayS/pgque#292
state=MERGED
draft=false
diff_lines=576
diff_added=264
diff_removed=100
diff_bytes=24765
comments_count=2
commits_count=1
ci_status=success
ci_summary=total=14 success=14 failure=0 pending=0 other=0
prompt=.claude/commands/review-mr.md
blocking=false
posted_by=gh
no_comment=false
live_posting=posted
```

</details>

---
*samorev-assisted review (AI analysis by [Tanya301/samorev](https://github.com/Tanya301/samorev))*
